### PR TITLE
[Enhancement] Show test names to students

### LINKED ISSUE #2359: Test names missing in student view
### Describe the bug

When student's check their test results, they cannot see the name of the test case, which makes the feedback pretty much useless. See here 
image

If an instructor uses the "Scores" view, however, they can see the name. See here 
 
image

#### To Reproduce

Submit anything that causes a failing test and check both views.

#### Expected behavior

The test name should be visible for students. Otherwise, they cannot know what parameters made which test case fail.

#### Screenshots

See above

### Environment

<details><pre>
</pre></details>

#### Output in Web Console

<details><pre>
  [REPLACE THIS WITH YOUR INFORMATION]
</pre></details>

### PR DESCRIPTION
**Pull request by schultek**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Resolves #2359 

### Description

- Add a new 'showTestNamesToStudents' property to the programming exercise entity
- Use this property in all places where the result-detail component is opened
  - When enabled, the student effectively sees the same information on feedbacks as the instructor
  - (The instructor always sees the test names, regardless of the option)
- Some refactoring for the modal creation to have proper type support.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Programming Exercise with the new option enabled
4. Participate as a student and view the feedback, test case names should be visible and test case feedback should be distinguishable from manual feedback.
5. Edit the exercise to disable the new option
6. View again as a student, there should not be any visual distinction between manual and automatic feedback or any test case names shown
7. View as an instructor, test case names should be visible and test case feedback should be distinguishable from manual feedback.
8. Delete the exercise.

### Test Coverage

**Client:**

### Screenshots

**Exercise Creation Option**
Bildschirmfoto 2020-11-12 um 16 40 57

**With option enabled** (Same as instructor view)
Bildschirmfoto 2020-11-12 um 16 48 52

**With option disabled** (Default)
Bildschirmfoto 2020-11-12 um 16 49 17

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 1]
```diff
@@ -66,6 +66,9 @@
     @Column(name = "sequential_test_runs")
     private Boolean sequentialTestRuns;
 
+    @Column(name = "show-test-names-to-students")
```
Comment (kloessst) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:69]: I think we use underscores normally

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-4.1.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd">
+    <changeSet author="schultek" id="20201111141743">
+        <addColumn tableName="programming_exercise_details">
+            <column name="show-test-names-to-students" type="BOOLEAN" />
```
Comment (kloessst) [inline src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml:5]: Use underscores for column name and defaultValueBoolean="false"

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 3]
```diff
@@ -44,6 +44,7 @@
             "onlineEditor": "Online Editor",
             "allowOfflineIde": "Offline IDE erlauben",
             "offlineIde": "Offline IDE",
+            "showTestNamesToStudents": "Zeige Testnamen an Studenten",
```
Comment (kloessst) [inline src/main/webapp/i18n/de/programmingExercise.json:47]:
```suggestion
            "showTestNamesToStudents": "Zeige Testnamen den Studenten",
```

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
```
Comment (kloessst) [inline src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml:1]: You need to include this file in `liquibase/master.xml`

**Review by kloessst — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java (hunk 1 above)
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java]: Ok 👍

File: src/main/webapp/i18n/de/programmingExercise.json (hunk 3 above)
Comment (schultek) [reply src/main/webapp/i18n/de/programmingExercise.json]: Done

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml (hunk 4 above)
Comment (schultek) [reply src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml:1]: Done

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml (hunk 2 above)
Comment (schultek) [reply src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml]: Done 👍

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 5]
```diff
@@ -66,6 +66,9 @@
     @Column(name = "sequential_test_runs")
     private Boolean sequentialTestRuns;
 
+    @Column(name = "show_test_names_to_students")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:69]:
```suggestion
    @Column(name = "show_test_names_to_students", table = "programming_exercise_details")
```

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-4.1.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd">
+    <changeSet author="schultek" id="20201111141743">
+        <addColumn tableName="programming_exercise_details">
+            <column name="show_test_names_to_students" type="BOOLEAN" defaultValueBoolean="false" />
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml:5]: not sure if BOOLEAN is the right type, please check the other xml files

File: src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java [hunk 7]
```diff
@@ -66,6 +66,9 @@
...
+    private Boolean showTestNamesToStudents;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:70]:
```suggestion
    private boolean showTestNamesToStudents;
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExercise.java:70]: this makes it easier if we have a default value

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 8]
```diff
@@ -44,6 +44,7 @@
             "onlineEditor": "Online Editor",
             "allowOfflineIde": "Offline IDE erlauben",
             "offlineIde": "Offline IDE",
+            "showTestNamesToStudents": "Zeige Testnamen den Studenten",
```
Comment (krusche) [inline src/main/webapp/i18n/de/programmingExercise.json:47]:
```suggestion
            "showTestNamesToStudents": "Zeige die Test Namen den Studierenden",
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html [hunk 9]
```diff
@@ -355,6 +355,21 @@ <h4 *ngIf="isImport" id="jhi-programming-exercise-heading-import" jhiTranslate="
                         </label>
                     </div>
                 </div>
+                <div class="form-group-narrow">
+                    <div class="form-check">
+                        <label class="form-check-label" for="field_showTestNamesToStudents">
```
Comment (krusche) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html:360]: can we have some explanation with a tooltip?

Comment (schultek) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.html:360]: yes, done

File: src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml (hunk 6 above)
Comment (schultek) [reply src/main/resources/config/liquibase/changelog/20201111141743_changelog.xml]: i changed this to 'bit' as seen in other changelogs

**Review by FrankeLukas — APPROVED** [review]
Works as advertised.

**Review by FuchsDominik — DISMISSED** [review]
Looks good, well done. At first i thought a instructor should also see the test names while participating, but then i realized this is only for assessing the exercise. Did not experience any weird behavior.

**Comment by balazs-czopf** [comment]
Looks good to me, nice work 🚀

**Review by balazs-czopf — DISMISSED** [review]
Looks good to me, nice work 🚀

**Review by krusche — DISMISSED** [review]

**Review by FrankeLukas — APPROVED** [review]
Re-approve.

**Review by krusche — APPROVED** [review]

**Pull request merged**